### ISSUE TITLE
warp: Incorrect handling of `If-None-Match` vs `If-Modified-Since`

### ISSUE THREAD

**Issue by brprice** [yesodweb/wai#842]
## In practice

I have seen a caching bug where I expected warp to serve a fresh version of a file, but firefox stubbonly used its stale cached copy.

I was running a local server serving some assets from `/nix/store` (where all the timestamps are set to the start of the unix epoch), with `Cache-Control: no-cache` (i.e. "verify with server before serving from local cache"). Load up the page in firefox, change the backing assets, rerun the server, load the page again in Firefox. I expected the ETag to change and thus invalidate the cache, but warp was replying 304, and so Firefox displayed a stale copy.

## Reproducer

Consider the following simple server using `warp` and `wai-app-static`:

```haskell
module Main where

import Network.Wai
import Network.Wai.Application.Static
import Network.Wai.Handler.Warp

app :: Application
app = staticApp $ defaultWebAppSettings "."

-- Add some logging, to diagnose the issue
logger :: Middleware
logger f req resp = do
  putStrLn "req hdrs:"
  print $ requestHeaders req
  f req $ \r -> do
    putStrLn "resp hdr:"
    print $ responseHeaders r
    putStrLn "resp status:"
    print $ responseStatus r
    resp r

main :: IO ()
main = run 8080 $ logger app
```

Assume we have a file `test` such that the response contains headers:
```
$ curl -s -o /dev/null -D - localhost:8080/test
HTTP/1.1 200 OK
Last-Modified: Sun, 25 Apr 2021 16:06:44 GMT
...
Server: Warp/3.3.15
...
ETag: 2Oj8otwPiW/Xy0ywAxuiSQ==
```

Now make a request with some caching headers. We will have an `If-Modified-Since` which matches the `Last-Modified`, but an `If-None-Match` that does **not** match the `ETag`:
```
$ curl -s -o /dev/null -D - localhost:8080/test -H 'If-Modified-Since: Sun, 25 Apr 2021 16:06:44 GMT' -H 'If-None-Match: AAAAAAAAAAAAAAAAAAAAAA=='
HTTP/1.1 304 Not Modified
Date: Sun, 25 Apr 2021 16:16:40 GMT
Server: Warp/3.3.15
Content-Type: application/octet-stream
Cache-Control: public, max-age=31536000
Expires: Thu, 31 Dec 2037 23:55:55 GMT
ETag: 2Oj8otwPiW/Xy0ywAxuiSQ==
```

Notice that this responds with `304 Not Modified`, which is wrong. We should return `200 OK`.

According to https://tools.ietf.org/html/rfc7232#section-3.3
> A recipient MUST ignore If-Modified-Since if the request contains an
   If-None-Match header field; the condition in If-None-Match is
   considered to be a more accurate replacement for the condition in
   If-Modified-Since...

[I think all other combinations of these two headers being absent/matching/not-matching are handled correctly]

## Details
If we look at the logging messages that warp spits out above, we see that
- wai-app-static deals with etags headers (producing `ETag` and matching `If-None-Match` to decide whether to reply `304`) and doesn't care about last-modified times (does not produce a `Last-Modified` header nor matches an `If-Modified-Since`)
- warp deals with last-modified times (producing `Last-Modified` header and matching an `If-Modified-Since`)
- if wai-app-static returns `304 Not Modified`, then warp will report that
- if wai-app-static returns `200 OK`, then warp will check the last-modified date compared to the `If-Modified-Since` and replies `304` if they match.

This has the unfortunate result that if the etags do not match, but the dates do, then wai-app-static will correctly reply with a `200`, but warp will ignore the etags, and incorrectly "upgrade" this to a `304`.

## Version information
- OS: NixOS 20.09
- cabal-install version 3.4.0.0
- The Glorious Glasgow Haskell Compilation System, version 8.10.4
-  wai == 3.2.3
-  wai-app-static == 3.1.7.2
- warp == 3.3.15

**Comment by brprice** [comment]
One workaround is to disable caching completely, though this obviously is not a good long-term solution. You can achieve this using the middleware
```haskell
noCache :: WAI.Middleware
noCache = WAI.modifyResponse $ WAI.mapResponseHeaders (("Cache-Control", "no-store") :)
```

**Comment by brprice** [comment]
I suspect this same issue is at play in #680 , which asks for a workaround of just not setting a `Last-Modified` header.

**Comment by snoyberg** [comment]
Issue #680 is a WONTFIX and I'm going to close it.

It seems like you're asking for logic added to warp that ignores if-modified-since if if-none-match is provided. Is that correct?

**Comment by kazu-yamamoto** [comment]
This cache logic of Warp was implemented by me. I decided to not use e-tag in the Warp level because calculating e-tag is costly. 

To skip the logic, Warp need to know the upper-level application used its own cache mechanism. I suggest to use `Vault` in `Request` since `conditionalRequest` is called in `sendRsp` called in `sendResponse` which takes `Request`. So, it's quit easy to pass this value to `conditionalRequest`.

Summary:
- Adding an API to ask Warp to skip its cache mechanism via `Vault`

What do you think?

**Comment by kazu-yamamoto** [comment]
@brprice Gentle ping.

**Comment by brprice** [comment]
Thanks for the ping!

@snoyberg 
> It seems like you're asking for logic added to warp that ignores if-modified-since if if-none-match is provided. Is that correct?

Yes, I think this would fix it. I think we want both that
- when receiving a request with both if-modified-since and if-none-match we should ignore if-modified-since
- when sending a response, do not add last-modified if a underlying application has added an etag, to avoid problems with browsers that do not support if-none-match (is this a concern these days? I am slightly concerned that we may serve both headers, and the an old browser ignore the etag and only give a if-modified-since on the next request.)

@kazu-yamamoto
> Adding an API to ask Warp to skip its cache mechanism via `Vault`

I don't know much about `Vault`, but it sounds like this could work. (It sounds like the same basic idea as @snoyberg, but with manual control, rather than by inspecting the headers?) I'm a bit concerned that this is another think one has to know to use, rather than having servant+wai-app-static caching working well out of the box.

**Comment by Vlix** [comment]
Is it safe to say, that:
- if there __IS__ an `ETag` header in the response headers, then an `If-None-Match` header should overrule any present `If-Modified-Since` header?
- if there's __NO__ `ETag` header in the response headers, then an `If-None-Match` header should be ignored and any present `If-Modified-Since` header should be used?

**Comment by brprice** [comment]
I'm afraid I don't understand that question -- `ETag`s are response headers and `If-*` are request headers; how does the response and request relate? Consider two request/response pairs for the same resource: `Client --[req1]--> Server --[resp1]--> Client --[req2]--> Server --[resp2]--> Client`, do you mean the situation considering
-  `ETag` in `resp1`, `If-*` headers in the later request `req2` (this would require the server to be more stateful)
- `ETag` in `resp1`, `If-*` headers in the request-we-are-currently-responding-to `req1` (this is temporally confusing to me)
- something else?

**Comment by Vlix** [comment]
Well, yes, `warp` knows what the headers in the response are going to be before it overrides the response to be 304. So if `req1` has an `If-*` header, we can check if any app or middleware has set an `ETag` header in the response to decide whether we want to ignore the `If-Modified-Since` header if there's ALSO an `If-None-Match`. Since `warp` doesn't know the internals of the app, this might be one way to know/guess that the app running on `warp` supports `If-None-Match`, so then we can prevent it from taking the `If-Modified-Since` into account.

Did that make any sense?

**Comment by brprice** [comment]
Ah thanks, yes that makes sense.

> Is it safe to say, that:
> 
>     * if there **IS** an `ETag` header in the response headers, then an `If-None-Match` header should overrule any present `If-Modified-Since` header?

Yes, I think so.
 
>     * if there's **NO** `ETag` header in the response headers, then an `If-None-Match` header should be ignored and any present `If-Modified-Since` header should be used?

I think this may be ok in practice, though my understanding is that the RFC requires ignoring the `If-Modified-Since` header in this case also. Are there any situations that makes this actually happen -- where would the tag in `If-None-Match` come from? I can only think of the case where the server is modified between two requests such that it sent an `ETag` in its first response but in the second response does not support them any longer -- doing caching in this situation is somewhat worrying! (Though I think this behaviour would solve my issue in practice, and seems better than the status quo)

**Comment by Vlix** [comment]
> Are there any situations that makes this actually happen -- where would the tag in `If-None-Match` come from?

Hmmm, good point. RFC 9110 indeed states:

"A recipient MUST ignore `If-Modified-Since` if the request contains an [`If-None-Match`](https://www.rfc-editor.org/rfc/rfc9110.html#field.if-none-match) header field"

That should be taken into account. :+1:

**Comment by Vlix** [comment]
Has been fixed in #945 and will be published in `warp-3.3.30`
